### ISSUE TITLE
Create script to check project files for missing data

### ISSUE THREAD

**Issue by cnk** [hackforla/website#386]
### Overview 

We need an easy way of finding out when data is missing from project cards so that we can present an uniform UX and so that functions that rely on this data can work across all projects.

### Action Items

- [X] Determine desired output and process to follow
- [X] Identify data types (links, leadership, description, etc)
- [ ] Identify which items projects might not have (e.g. live site)
- [ ] Propose value to use for fields where data is intentionally missing 

### Requirements 

- Frequency: 1x/week 
- Output: ?

**Comment by zempo** [comment]
I will make a test repo with mock data to try some solutions 
It will expand upon @cnk 's previous [attempt](https://github.com/cnk/actions-test) to crawl through the project's directory.

- [ ] research options
- [ ] make test repo
- [ ] check with Bonnie

**Comment by rblaz001** [comment]
I'm just getting over some pretty bad food poisoning so I haven't worked on this as much as I would have liked. I still need to make a 100 Automation's repo to continue developing my python script.

**Action Item Progress**
- [x] Determine desired output and process to follow
- [x] Identify data types (links, leadership, description, etc)
- [ ] Identify which items projects might not have (e.g. live site)
- [ ] Propose value to use for fields where data is intentionally missing

I have chosen to stick to a spreadsheet for the desired output of our script. I'm currently outputting a spreadsheet that contains two columns, one for the name of the project and the other for the list of missing data types.

The data types that we are using are currently hard coded in the script and were obtained by using a script that collected a set of used data types across all projects. 

**Set of Data Types**
![Screenshot from 2020-08-01 14-23-59](https://user-images.githubusercontent.com/35281980/89110772-e8dc5b80-d402-11ea-9a0f-00b843673418.png)

**Project Missing Data Types**
![Screenshot from 2020-08-01 14-23-26](https://user-images.githubusercontent.com/35281980/89110769-e67a0180-d402-11ea-99d3-71eca9d65b6c.png)

**Note:** I'm currently not taking account of nested data types. For example name, role, and links that are under "Leadership" in the following image. This is something I want to look in to, but I might have to reconsider how I deal with my output and determine what nested data types are important.

![Screenshot from 2020-08-01 14-43-56](https://user-images.githubusercontent.com/35281980/89110991-85076200-d405-11ea-8114-321aafeecb2b.png)

**Comment by rblaz001** [comment]
I now have a version of the script that outputs unused data types and missing data types for all projects.
I still need create a 100 automation repo and add my work to it.
Remaining work is quality of life edits to the script, including proper command line support.

The data types that we are using are currently hard coded in the script and were obtained by using a script that collected a set of used data types across all projects. The nested data types are denoted by comma delimited entries.

**Set of Data Types**
![Screenshot from 2020-08-09 09-56-14](https://user-images.githubusercontent.com/35281980/89737587-dd062000-da26-11ea-87b9-c3b0760a22b4.png)

**Project Unused Data Types**
![Screenshot from 2020-08-09 09-55-02](https://user-images.githubusercontent.com/35281980/89737592-e2fc0100-da26-11ea-804c-4e551fb35353.png)

**Project Missing Data Types (Currently Only One Missing)**
![Screenshot from 2020-08-09 09-55-21](https://user-images.githubusercontent.com/35281980/89737595-e7281e80-da26-11ea-8e33-77775f6ba6e5.png)
- need to change output to properly where the missing data is nested.

**Example of role missing from above entry**
![Screenshot from 2020-08-09 09-58-00](https://user-images.githubusercontent.com/35281980/89737599-ebecd280-da26-11ea-88c3-4cf6c0d85ced.png)

**Comment by rblaz001** [comment]
Added 100Automations repo. https://github.com/100Automations/script-gather-missing-data-types

**Comment by alexandrastubbs** [comment]
@ExperimentsInHonesty will talk to @rblaz001 about topic tags

**Comment by ExperimentsInHonesty** [comment]
@rblaz001  Please fill this out, so that we can get you listed on the 100automations.org site that will soon exist:
[create a project card for a 100 automations project](https://github.com/100Automations/Website/issues/new?assignees=&labels=documentation%2C+good+first+issue&template=create-automation-card-for--automation-name-.md&title=Create+automation+card+for+%5Bautomation+name%5D)

**Comment by rblaz001** [comment]
Spreadsheet for current unused/missing data types.

Note: Unused Data Types are data types that are never used in a project card. Missing Data Types are nested data types that are missing even though the parent data types are present. (This can be one or more instances)

[missingDataTypes.xlsx](https://github.com/hackforla/website/files/5081019/missingDataTypes.xlsx)

**Comment by rblaz001** [comment]
Notes From Bonnie:
- [x] Research Github Action integration for creating issues
- [x] Add new validation method. Provide a source .md collection file with every data type in project with a  require value or ignore value. This will be used to determine how script parses missing data types.
- [x] Add functionality to output report of missing data types of source .md file. This will make sure that source .md file is kept up to date with requirements. 
- [x] Add additional output option as JSON
- [ ] Add Github Action integration

**Comment by alexandrastubbs** [comment]
@rblaz001 do you have a list of all the data types? If so, can you upload to google drive and share the link here?

**Comment by rblaz001** [comment]
@alexandrastubbs I have to make a small adjustment to the script to output all the currently used data types. I'm not home at the moment but I'll be able to upload the file and link it before tomorrow morning.

**Comment by rblaz001** [comment]
@alexandrastubbs I added an extra sheet labeled "Used Data Types" that has a list of data types used across the project cards.

https://docs.google.com/spreadsheets/d/1w59UfwbGiEAoBXjAxl4uwQe3xk5j6yFd3NxBMOm510Y/edit?usp=sharing

Note: The data types are not ordered. Also comma separated data types represent the nesting
For example One,Two,Three would look like the following in the project card
```
One:
    Two:
        Three: value
```

**Comment by rblaz001** [comment]
**Progress Report**
Progress - Was able to finish refactor, add JSON support, and update README. I was also able to update repo name and logo to include Jekyll
Blockers - Need to show progress to PM and get notes
Availability: 3-4 days depending on possible job interviews
ETA:  
- Automation Card: End of week at most but shooting for much sooner
- Validate Through Template Feature: 7-9 days, difficulty level is relatively low but I'm catching up on documentation from refactor plus food-oasis tasks.
- GitHub Action Issues: TBD, began research 

**Work In Progress**
Progress with automation card:
https://github.com/100Automations/Website/issues/29

Finishing refactor documentation:
https://github.com/100Automations/jekyll-gather-data-types/issues/2

Then working on validate through template feature:
https://github.com/100Automations/jekyll-gather-data-types/issues/3

Was also able to look in to github action api and it should be possible to automate issues for missing data types relatively easy. The only blocker I'm having right now is thinking through how to manage issues so redundant issues are not created.

**Comment by alexandrastubbs** [comment]
@rblaz001 to make a tab with all the data types (missing, unused, used)

**Comment by alexandrastubbs** [comment]
@rblaz001 to rename from 'used data types' to 'all data types'

**Comment by rblaz001** [comment]
Progress - My availability changed this last week, I was unable to get any work done
Blockers - None
Availability: No more interviews or scheduling conflicts, will be available all week

Priority:
 - [x] Finish automation card
 - [x] Rename 'used data types' to 'all data types'
 - [x] Add validate through template feature

Task after priority finished:
 - [ ] GitHub actions for automatic issue creation

**Comment by ExperimentsInHonesty** [comment]
@rblaz001 checking to see if you are able to put in an update on this issue before tomorrow.

**Comment by rblaz001** [comment]
**Progress** 
- Finished [automation card](https://github.com/100Automations/Website/issues/29), renamed 'used data types' to 'all data types ' in output, and finished updating documentation.

Next step is to add validate through template. 
 - ~~I will work on this tonight and either should finish first draft of it tonight, or will update with expected completion time.~~
 - Finished MVP for validate through template, still need to update documentation and README before merging to master.

Then begin work on implementing GitHub actions

Blockers - Moving to Austin, still trying to find an apartment so I have very limited time

Availability - Limited

**Comment by ExperimentsInHonesty** [comment]
Do UX requirement gathering.
Bonnie will post on CfA slack asking for other people who have Jekyll sites and use collections.

**Comment by rblaz001** [comment]
**template.md**
```
---
alt: important
alt-hero: important
completed-contact: important
description: important
hide: ignore
identification: important
image: important
image-hero: important
leadership:
  - links:
      github: important
      linkedin: important
      slack: important
    name: important
    picture: important
    role: important
links:
  - name: important
    url: important
location: important
looking:
  - category: important
    skill: important
partner: important
status: important
technologies: important
title: important
tools: important
---
```

**Comment by rblaz001** [comment]
**Updated Project Card Data Types** - 9/13/2020
https://docs.google.com/spreadsheets/d/1F8ba2Kn6fjpp15zpseVqoYP7DsWVq1tpsKPusMCHF5Q/edit?usp=sharing

**Updated Template Data Types** - 9/13/2020
https://docs.google.com/spreadsheets/d/1KqqzZJOZKhX5QAB9X1wOwn_RuovRH1ikGwnQdUudbaM/edit?usp=sharing

**Comment by rblaz001** [comment]
Progress: Still have Validate by Template in Develop branch until I finish updating documentation. I hope to finish updating the README and pushing this to master today. Won't be able to make it to meeting today.

Blockers: Moving in a week

Availability: Today (Sunday 9/20) - I'm moving in a week, won't expect to work on anything until I'm settled in to my new place

**Comment by rblaz001** [comment]
Updated the README and merged Develop branch to master. I'll be driving to Austin starting Monday 9/28 and should be in my new apartment Thursday 10/1.

**Comment by rblaz001** [comment]
**Progress**
Starting work on GitHub Actions for Jekyll Gather Data Types. 

Still don't have much feedback on what a user expects for GitHub Action related features but will start by having issue generation based off of verify by template. Need to figure out a system that will track issues generated by Jekyll Gather Data Types to prevent redundant issue generation.

Blockers: Starting work and 10/12 and setting up a development environment on a new computer.

Availability: M-F 6:00 - 9:00 pm CT

**Comment by rblaz001** [comment]
No progress to report. Worked on wins page and previous status and availability still stands.

**Comment by ExperimentsInHonesty** [comment]
@rblaz001 checking in on you.  Let's catch up on what you need from me on this issue.

**Comment by ahdithebomb** [comment]
@ExperimentsInHonesty messaged @rblaz001 the following via slack. 

> Hi Raudel hope everything is going well on your new job.  Do you want us to hand this issue off to someone else, and if so, is there any information you need to impart before we do?

**Comment by ExperimentsInHonesty** [comment]
Review the feasibility of the following - Editing the script so that it produces the spreadsheet (instead of it being made manually) and another tab on the sheet that has only what project should have but is missing.  In order to determine that, use the criteria below:


**Current projects must have:**
```yaml
identification: ''
title:
description:
image: /assets/images/projects/[project repo name + .jpg or .png]
alt: 'description of the card image'
image-hero: /assets/images/projects/[project repo name + -hero + .jpg or .png]
leadership:
  - name: 
    role: 
    links:
      slack: 'https://hackforla.slack.com/team/xxxxxx'
      github: 'https://github.com/githubusername'
    picture: https://avatars.githubusercontent.com/githubusername
links: 
  - name: Github
    url: 'https://www.example.com'
  - name: Slack
    url: 'https://www.example.com'
  - name: Test Site
    url: 'https://test.example.com'
  - name: Demo Site
    url: 'https://demo.example.com'
  - name: Site
    url: 'https://live.example.com'
  - name: Overview
    url: https://github.com/hackforla/product-management/blob/master/project-one-sheets/[REPLACE WITH PROJECT NAME]-Project-One-Sheet.pdf
  # unused links can be commented out
  # - name: Showcase deck
  #   url: ''
  #   alt: ''
looking:
  - category: Development
    skill: one skill
  - category: UI/UX
    skill: another skill
  - category: Content
    skill: Researcher
technologies: 
  - Node.js 
  - ReactJS 
  - Ruby on Rails
  - other etc.
location: 
  - Remote
  - Downtown LA
  - Santa Monica
  - South LA
partner:
tools: figma, photoshop, sketch, phone calls.
vertical:
status: Active
```

**Completed projects must have** 
```
identification: ''
title:
description:
image: /assets/images/projects/[project repo name + .jpg or .png]
alt: 'description of the card image'
links: 
  - name: Github
    url: 'https://www.example.com'
  - name: Test Site
    url: 'https://test.example.com'
  - name: Demo Site
    url: 'https://demo.example.com'
  - name: Site
    url: 'https://live.example.com'
  - name: Overview
    url: https://github.com/hackforla/product-management/blob/master/project-one-sheets/[REPLACE WITH PROJECT NAME]-Project-One-Sheet.pdf
  # unused links can be commented out
  # - name: Showcase deck
  #   url: ''
  #   alt: ''
technologies: 
  - Node.js 
  - ReactJS 
  - Ruby on Rails
  - other etc.
partner:
vertical:
status: Completed
```
**_they can also have (but not mandatory)_**
```
image-hero: /assets/images/projects/[project repo name + -hero + .jpg or .png]
leadership:
  - name: 
    role: 
    links:
      slack: 'https://hackforla.slack.com/team/xxxxxx'
      github: 'https://github.com/githubusername'
    picture: https://avatars.githubusercontent.com/githubusername
```

**On Hold projects** 
have everything that Active projects have except 
No Looking for
Status: On Hold

**Comment by ExperimentsInHonesty** [comment]
@rblaz001 - sent you this on slack.  Realized you might not check slack so sending it here too

Hi.  Recently I made an update to your issue re the gather data types.  I then unassigned you intending to assign a junior person to take a look at the feasibility.  But I noticed today that the spreadsheet in the issue from the last time you produced a spreadsheet was now in the trash.  Which makes me think you might have picked it up again.

So now I realize I should have started by asking you.  Do you want some additional labor on this to produce the output (spreadsheet) we need to be able to use the scripts results on a weekly basis. I have a new backend dev that needs something meaty to work on.

If no, I will find something else for him

**Comment by ExperimentsInHonesty** [comment]
Also, I have suggested he look at the code you wrote for 100automations and see if he can do some of the things you asked for there (refactoring, etc).

**Comment by ExperimentsInHonesty** [comment]
sent raudel and email

**Comment by akibrhast** [comment]
Which fields should be allowed to be blank could you please specify using they keywords,
REQUIRED and OPTIONAL like in the example below.
@ExperimentsInHonesty @rblaz001

1. Which Locations Should be Allowed? Should it be allowed to have multiple locations
 - ["Remote","Downtown LA","Santa Monica","South LA","Los Angeles"]

```
identification: ''
title: REQUIRED
description: OPTIONAL
image: /assets/images/projects/[project repo name + .jpg or .png]
alt: 'description of the card image'
image-hero: /assets/images/projects/[project repo name + -hero + .jpg or .png]
leadership:
  - name: REQUIRED
    role: REQUIRED
    links: REQUIRED
      slack: 'https://hackforla.slack.com/team/xxxxxx'
      github: 'https://github.com/githubusername'
    picture: https://avatars.githubusercontent.com/githubusername
links: 
  - name: Github
    url: 'https://www.example.com'
  - name: Slack
    url: 'https://www.example.com'
  - name: Test Site
    url: 'https://test.example.com'
  - name: Demo Site
    url: 'https://demo.example.com'
  - name: Site
    url: 'https://live.example.com'
  - name: Overview
    url: https://github.com/hackforla/product-management/blob/master/project-one-sheets/[REPLACE WITH PROJECT NAME]-Project-One-Sheet.pdf
  # unused links can be commented out
  # - name: Showcase deck
  #   url: ''
  #   alt: ''
looking:
  - category: Development
    skill: one skill
  - category: UI/UX
    skill: another skill
  - category: Content
    skill: Researcher
technologies: 
  - Node.js 
  - ReactJS 
  - Ruby on Rails
  - other etc.
location: 
  - Remote
  - Downtown LA
  - Santa Monica
  - South LA
partner:
tools: figma, photoshop, sketch, phone calls.
vertical:
status: Active
```

**Comment by akibrhast** [comment]
@rblaz001 @ExperimentsInHonesty I generated this following the requirements listed above. Is this the type of report you are expecting/wanting? Using the comment from @bonnie -> https://github.com/hackforla/website/issues/386#issuecomment-785427876 as the source of truth

<details>
<summary>Active Projects Validation Results</summary>

```json
{
    "filename": "100-automations.md",
    "errors": [
        {
            "message": "requires property \"partner\"",
            "stack": "instance requires property \"partner\""
        },
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        }
    ]
}{
    "filename": "civic-tech-index.md",
    "errors": [
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        }
    ]
}{
    "filename": "brigade-organizers-playbook.md",
    "errors": [
        {
            "message": "requires property \"vertical\"",
            "stack": "instance requires property \"vertical\""
        }
    ]
}{
    "filename": "ballot-nav.md",
    "errors": [
        {
            "message": "is not of a type(s) string",
            "stack": "instance.leadership[0].role is not of a type(s) string"
        },
        {
            "message": "requires property \"technologies\"",
            "stack": "instance requires property \"technologies\""
        },
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        }
    ]
}{
    "filename": "civic-tech-structure.md",
    "errors": [
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        }
    ]
}{
    "filename": "311-data.md",
    "errors": [
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        }
    ]
}{
    "filename": "food-oasis.md",
    "errors": [
        {
            "message": "is not of a type(s) string",
            "stack": "instance.leadership[4].role is not of a type(s) string"
        }
    ]
}{
    "filename": "civic-opportunity-project.md",
    "errors": [
        {
            "message": "requires property \"technologies\"",
            "stack": "instance requires property \"technologies\""
        },
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        }
    ]
}{
    "filename": "engage.md",
    "errors": [
        {
            "message": "is not of a type(s) string",
            "stack": "instance.leadership[0].role is not of a type(s) string"
        }
    ]
}{
    "filename": "open-community-survey.md",
    "errors": [
        {
            "message": "requires property \"technologies\"",
            "stack": "instance requires property \"technologies\""
        }
    ]
}{
    "filename": "undebate.md",
    "errors": [
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        }
    ]
}{
    "filename": "tdm-calculator.md",
    "errors": [
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        }
    ]
}{
    "filename": "youthjusticenav.md",
    "errors": [
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        }
    ]
}{
    "filename": "writeforall.md",
    "errors": [
        {
            "message": "requires property \"technologies\"",
            "stack": "instance requires property \"technologies\""
        },
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        }
    ]
}{
    "filename": "vrms.md",
    "errors": [
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        }
    ]
}{
    "filename": "website.md",
    "errors": [
        {
            "message": "requires property \"partner\"",
            "stack": "instance requires property \"partner\""
        }
    ]
}
```

</details>

<details>
<summary>Completed Projects Validation Result</summary>

```json
{
    "filename": "heart.md",
    "errors": [
        {
            "message": "requires property \"github\"",
            "stack": "instance.leadership[0].links requires property \"github\""
        },
        {
            "message": "requires property \"slack\"",
            "stack": "instance.leadership[0].links requires property \"slack\""
        },
        {
            "message": "is not allowed to have the additional property \"linkedin\"",
            "stack": "instance.leadership[0].links is not allowed to have the additional property \"linkedin\""
        },
        {
            "message": "requires property \"github\"",
            "stack": "instance.leadership[1].links requires property \"github\""
        },
        {
            "message": "requires property \"slack\"",
            "stack": "instance.leadership[1].links requires property \"slack\""
        },
        {
            "message": "is not allowed to have the additional property \"linkedin\"",
            "stack": "instance.leadership[1].links is not allowed to have the additional property \"linkedin\""
        },
        {
            "message": "requires property \"github\"",
            "stack": "instance.leadership[2].links requires property \"github\""
        },
        {
            "message": "requires property \"slack\"",
            "stack": "instance.leadership[2].links requires property \"slack\""
        },
        {
            "message": "is not allowed to have the additional property \"linkedin\"",
            "stack": "instance.leadership[2].links is not allowed to have the additional property \"linkedin\""
        },
        {
            "message": "requires property \"slack\"",
            "stack": "instance.leadership[3].links requires property \"slack\""
        },
        {
            "message": "is not allowed to have the additional property \"linkedin\"",
            "stack": "instance.leadership[3].links is not allowed to have the additional property \"linkedin\""
        },
        {
            "message": "requires property \"slack\"",
            "stack": "instance.leadership[4].links requires property \"slack\""
        },
        {
            "message": "is not allowed to have the additional property \"linkedin\"",
            "stack": "instance.leadership[4].links is not allowed to have the additional property \"linkedin\""
        },
        {
            "message": "requires property \"slack\"",
            "stack": "instance.leadership[5].links requires property \"slack\""
        },
        {
            "message": "is not allowed to have the additional property \"linkedin\"",
            "stack": "instance.leadership[5].links is not allowed to have the additional property \"linkedin\""
        },
        {
            "message": "requires property \"slack\"",
            "stack": "instance.leadership[6].links requires property \"slack\""
        },
        {
            "message": "is not allowed to have the additional property \"linkedin\"",
            "stack": "instance.leadership[6].links is not allowed to have the additional property \"linkedin\""
        },
        {
            "message": "requires property \"slack\"",
            "stack": "instance.leadership[7].links requires property \"slack\""
        },
        {
            "message": "is not allowed to have the additional property \"linkedin\"",
            "stack": "instance.leadership[7].links is not allowed to have the additional property \"linkedin\""
        },
        {
            "message": "requires property \"slack\"",
            "stack": "instance.leadership[8].links requires property \"slack\""
        },
        {
            "message": "is not allowed to have the additional property \"linkedin\"",
            "stack": "instance.leadership[8].links is not allowed to have the additional property \"linkedin\""
        },
        {
            "message": "requires property \"technologies\"",
            "stack": "instance requires property \"technologies\""
        }
    ]
}{
    "filename": "hellogov.md",
    "errors": [
        {
            "message": "requires property \"role\"",
            "stack": "instance.leadership[0] requires property \"role\""
        },
        {
            "message": "requires property \"role\"",
            "stack": "instance.leadership[1] requires property \"role\""
        },
        {
            "message": "requires property \"role\"",
            "stack": "instance.leadership[2] requires property \"role\""
        },
        {
            "message": "requires property \"technologies\"",
            "stack": "instance requires property \"technologies\""
        }
    ]
}{
    "filename": "metro-ontime.md",
    "errors": [
        {
            "message": "requires property \"technologies\"",
            "stack": "instance requires property \"technologies\""
        }
    ]
}{
    "filename": "work-for-la.md",
    "errors": [
        {
            "message": "requires property \"slack\"",
            "stack": "instance.leadership[0].links requires property \"slack\""
        },
        {
            "message": "requires property \"technologies\"",
            "stack": "instance requires property \"technologies\""
        }
    ]
}
```
</details>

<details>
<summary>On Hold Projects Validation Result</summary>

```json
{
    "filename": "adopt-civic-art.md",
    "errors": [
        {
            "message": "requires property \"image-hero\"",
            "stack": "instance requires property \"image-hero\""
        },
        {
            "message": "requires property \"leadership\"",
            "stack": "instance requires property \"leadership\""
        },
        {
            "message": "requires property \"technologies\"",
            "stack": "instance requires property \"technologies\""
        },
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        },
        {
            "message": "requires property \"vertical\"",
            "stack": "instance requires property \"vertical\""
        }
    ]
}{
    "filename": "criminal-sentencing.md",
    "errors": [
        {
            "message": "requires property \"image-hero\"",
            "stack": "instance requires property \"image-hero\""
        },
        {
            "message": "requires property \"leadership\"",
            "stack": "instance requires property \"leadership\""
        },
        {
            "message": "requires property \"technologies\"",
            "stack": "instance requires property \"technologies\""
        },
        {
            "message": "requires property \"partner\"",
            "stack": "instance requires property \"partner\""
        },
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        },
        {
            "message": "requires property \"vertical\"",
            "stack": "instance requires property \"vertical\""
        }
    ]
}{
    "filename": "curbmap.md",
    "errors": [
        {
            "message": "requires property \"identification\"",
            "stack": "instance requires property \"identification\""
        },
        {
            "message": "requires property \"image-hero\"",
            "stack": "instance requires property \"image-hero\""
        },
        {
            "message": "requires property \"leadership\"",
            "stack": "instance requires property \"leadership\""
        },
        {
            "message": "requires property \"technologies\"",
            "stack": "instance requires property \"technologies\""
        },
        {
            "message": "requires property \"partner\"",
            "stack": "instance requires property \"partner\""
        },
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        },
        {
            "message": "requires property \"vertical\"",
            "stack": "instance requires property \"vertical\""
        }
    ]
}{
    "filename": "light-the-way.md",
    "errors": [
        {
            "message": "requires property \"identification\"",
            "stack": "instance requires property \"identification\""
        },
        {
            "message": "requires property \"image-hero\"",
            "stack": "instance requires property \"image-hero\""
        },
        {
            "message": "requires property \"leadership\"",
            "stack": "instance requires property \"leadership\""
        },
        {
            "message": "requires property \"technologies\"",
            "stack": "instance requires property \"technologies\""
        },
        {
            "message": "requires property \"partner\"",
            "stack": "instance requires property \"partner\""
        },
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        },
        {
            "message": "requires property \"vertical\"",
            "stack": "instance requires property \"vertical\""
        }
    ]
}{
    "filename": "jobs-for-hope.md",
    "errors": [
        {
            "message": "requires property \"leadership\"",
            "stack": "instance requires property \"leadership\""
        },
        {
            "message": "requires property \"technologies\"",
            "stack": "instance requires property \"technologies\""
        },
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        },
        {
            "message": "requires property \"vertical\"",
            "stack": "instance requires property \"vertical\""
        }
    ]
}{
    "filename": "new-schools-today.md",
    "errors": [
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        },
        {
            "message": "requires property \"vertical\"",
            "stack": "instance requires property \"vertical\""
        }
    ]
}{
    "filename": "spare.md",
    "errors": [
        {
            "message": "requires property \"leadership\"",
            "stack": "instance requires property \"leadership\""
        },
        {
            "message": "requires property \"technologies\"",
            "stack": "instance requires property \"technologies\""
        },
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        },
        {
            "message": "requires property \"vertical\"",
            "stack": "instance requires property \"vertical\""
        }
    ]
}{
    "filename": "shared-housing-project.md",
    "errors": [
        {
            "message": "requires property \"leadership\"",
            "stack": "instance requires property \"leadership\""
        },
        {
            "message": "requires property \"technologies\"",
            "stack": "instance requires property \"technologies\""
        },
        {
            "message": "requires property \"partner\"",
            "stack": "instance requires property \"partner\""
        },
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        },
        {
            "message": "requires property \"vertical\"",
            "stack": "instance requires property \"vertical\""
        }
    ]
}{
    "filename": "not-today.md",
    "errors": [
        {
            "message": "requires property \"tools\"",
            "stack": "instance requires property \"tools\""
        }
    ]
}
```
</details>

**Comment by cnk** [comment]
What are the items like: 

```
{
    "filename": "food-oasis.md",
    "errors": [
        "is not of a type(s) string",
        "is not of a type(s) array"
    ]
}
```

**Comment by akibrhast** [comment]
ah.. my bad... let me work on a bit more detailed error report.
But in answer to your question. The two errors in question are happening because these fields['looking','role'] in the image below are empt/null.
Where my json schema is expecting them to be a string. @cnk 


![image](https://user-images.githubusercontent.com/32349001/110046175-8b521980-7d19-11eb-83de-89a2e11018fc.png)
![image](https://user-images.githubusercontent.com/32349001/110046221-a2910700-7d19-11eb-9421-9cd47d57ad0e.png)

**Comment by akibrhast** [comment]
Edited https://github.com/hackforla/website/issues/386#issuecomment-790993825 , to give a bit more detailed report.
I was really hoping to get something along the line of
```json
"error":[
 {
 "message": "is not of type string",
 "expected": "string",
 "given": "null",
 "stack": "on line 256 ,in 'message': null"
 }
]

```
but can't seem to find the relevant information... :(

**Comment by cnk** [comment]
"looking" (and probably "technologies") can be empty. Can you express that in a json schema? optional fields can exist but be empty?

**Comment by akibrhast** [comment]
yewp, not a problem. I can make the looking be either string/null . But a project should always at least have technology, no?? I would be hesitant on allowing that to be empty...

**Comment by akibrhast** [comment]
Updated - >  https://github.com/hackforla/website/issues/386#issuecomment-790993825
An empty looking field does not report an error. Although it is still a required key.

**Comment by akibrhast** [comment]
Currently I have a `on: workflow dispatch` action running that checks the files and generates a validation report. @ruben1s 

- https://github.com/akibrhast/website/actions/runs/652747129
- https://github.com/akibrhast/website/blob/gh-pages/_data/project_validation_report.json


Require feedback on where you want to go next with this @cnk? @ExperimentsInHonesty ?

My Proposal
---
- Have this github action run ever 3.5days
- On action start
  1. Check and see if an issue created by the action already exists
  2. If an open issue exists that was created by this bot-action
        - Quit Action
> This should be fairly easy to implement using this endpoint <pre> <a  href="https://api.github.com/repos/hackforla/website/issues?state=open&creator=akibrhast">https://api.github.com/repos/hackforla/website/issues?state=open&creator=akibrhast</a></pre>
  3. If an open issue does not exist that was created by this bot-action
        - Generate validation report and create an issue with the report data

<br>
Example Issue that would be created : 
<br>

## Project Validation Failing[month/date/year]

### Overview
The following report was generated due to project schema not matching validation schema. Please review the report listed below under `Project Validation Error Report` and apply the necessary fixes.

<details>
<summary>Project Validation Error Report</summary>

```json
[
    {
        "file": "100 Automations",
        "errors": [
            {
                "message": "requires property \"partner\"",
                "stack": "instance requires property \"partner\""
            },
            {
                "message": "requires property \"tools\"",
                "stack": "instance requires property \"tools\""
            }
        ]
    },
    {
        "file": "311 Data",
        "errors": [
            {
                "message": "requires property \"tools\"",
                "stack": "instance requires property \"tools\""
            }
        ]
    }
]
```

</details>

## Action Items
- [ ] Please apply the fixes to the files listed in the resources below using the error report above

## Resources/Instructions
- https://github.com/hackforla/website/blob/gh-pages/_projects/100-automations.md
- https://github.com/hackforla/website/blob/gh-pages/_projects/311-data.md

**Comment by ExperimentsInHonesty** [comment]
@akibrhast I like the proposal.  

- Still trying to figure out how this would be used. 
  - concerned that if it has one issue that has a lot of different projects in it, it will be hard to parse out.  We can't get 10 PMs to weigh in on one issue.
  - concerned about us determining that a project was not going to have something in it (like no partner) and no way to get it to stop flagging it.
 
Interested to hear your suggestions for improvement based on these concerns.

**Comment by akibrhast** [comment]
> concerned about us determining that a project was not going to have something in it (like no partner) and no way to get it to stop flagging it.

Currently the schema is created based on this comment  https://github.com/hackforla/website/issues/386#issuecomment-785427876 ('Current Project Must Have's') .

According to that comment, all current projects must have those keys. However it seems like if you look at the image below that key is commented out for some reason. https://raw.githubusercontent.com/hackforla/website/gh-pages/_projects/100-automations.md . That should not be the case, it's fine if the partner value is empty, but key should still exist.

![image](https://user-images.githubusercontent.com/32349001/111254384-51bbc100-85eb-11eb-8e52-3965dc093308.png)

### According to the original issue  Action Item https://github.com/hackforla/website/issues/386#issue-585777689
- [ ] Identify which items projects might not have (e.g. live site)

#### Some details that needs to be consolidated are
- what are the keys that a current project MUST ALWAYS have
- what are the values that a current project MUST ALWAYS have (example: title,id)
- what are the values that current project may be OPTIONAL(ex: partners)
- what are the type of the value a field accepts for example a slack link should always look like `https://hackforla.slack.com/team/xxxxxx` .

Same questions apply to Completed Projects and On Hold Projects. 

Changing the schema is relatively simple. As you can see, a comment by @cnk https://github.com/hackforla/website/issues/386#issuecomment-791064664 was made and the schema was soon changed after that.

---

> concerned that if it has one issue that has a lot of different projects in it, it will be hard to parse out. We can't get 10 PMs to weigh in on one issue.

Maybe generate a single issue per file ?

**Comment by akibrhast** [comment]
## Progress
None since 5 days ago 
https://github.com/hackforla/website/issues/386#issuecomment-799935288

## Blockers
Awaiting Feedback from @ExperimentsInHonesty @cnk 

## Availability
Here and There

## ETA
N/A

**Comment by ExperimentsInHonesty** [comment]
Met with @akibrhast - he will add notes to this doc.
